Properly support SVGs

graphiteisaac opened the pull request:
This is a pretty simple, rudimentary addition to properly support Lustre SVG elements. There will inevitably be edge cases in the SVG spec I haven't considered, but for 99% of SVGs used on the web, this should work just fine, and be a nice helpful tool for people.

I added a couple of functions which pass an "is_svg" boolean down the chain, as it's helpful to consider context for elements, and considered a couple of edge cases in the attribute printer (eg, svg can't set the width with attribute.width, viewBox is case sensitive). There is one new test which covers both these cases, proving it works :)

graphiteisaac committed:
(days old, whoops) basic SVG support

graphiteisaac committed:
double compute attributes (sorry!) to make svgs work properly and easily

graphiteisaac committed:
add test, fix viewBox case

lpil commented on app/src/app.gleam:
```diff
@@ -19,9 +20,9 @@ pub type Model {
+const copy_button_default_text = "Copy"
 
-const copy_button_copied_text = "Copied 🥰"
+const copy_button_copied_text = "Copied"
```
Why have these been removed?

graphiteisaac replied:
I put in an SVG icon for the copy, which is perhaps a bit less joyous but is a bit more semantically clear

lpil replied:
One change per PR please, revert this

lpil commented on src/html_lustre_converter.gleam:
````diff
@@ -14,12 +15,19 @@ import javascript_dom_parser.{type HtmlNode, Comment, Element, Text} as parser
+/// If the source document contains SVGs, we need one more import:
+/// ```gleam
+/// import lustre/element/svg
+/// // or 
````
or?

lpil commented on src/html_lustre_converter.gleam:
```diff
@@ -184,14 +295,26 @@ fn print_element(
     }
 
+    "svg" -> {
+      // TODO: Fix double compute, avoid initial compute if we don't go to SVG land
```
What does this mean?

graphiteisaac replied:
Currently, we already compute the attributes at the start of the function, but then when in svg we compute them a second time with the svg-friendly output. I ended up leaving it in when I found no meaningful performance impact, I found it was not necessarily needed. Should have preferenced speaking about it in the PR over leaving a dud TODO :P

lpil replied:
Can just remove the comment I think

lpil commented on src/html_lustre_converter.gleam:
```diff
@@ -224,10 +347,16 @@ fn get_text_content(nodes: List(HtmlNode)) -> String {
 fn print_children(
   children: List(HtmlNode),
   ws: WhitespaceMode,
+  is_svg: Bool,
```
Prefer a custom type to a bool 🙏

graphiteisaac replied:
Will update :)

lpil reviewed (commented):
Ahh thank you!! I've left a few notes

graphiteisaac committed:
Fixes from PR feedback

Includes moving to a custom OutputMode type rather than booleans,
cleaning up some features I hadn't finished to the degree I'm
comfortable with, and removing a TODO that was not necessary.

graphiteisaac committed:
Format test file properly

graphiteisaac committed:
revert frontend changes - keep for other PR

graphiteisaac commented:
Thanks very much Louis :) I've made numerous changes based on your feedback now.

lpil reviewed (approved):
thank you!!!!!
